### PR TITLE
Valhalla: value type cast to null changes to checkcast

### PR DESCRIPTION
**Pull request by XxAdi101xX**
These changes introduce functionality that throws a null pointer exception if a checkcast detects that a null reference is being casted to a value type. Tests for this are also included here.

### CODE REVIEW

**Comment by XxAdi101xX** [comment]
@tajila pls take a look when you have a chance

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 1]
```diff
@@ -1440,6 +1440,14 @@ static public void testDefaultValueInassortedRefWithLongAlignmentArray() throws
 		}
 	}
 
+	@Test(priority=4, expectedExceptions=NullPointerException.class)
+	static public void testCheckCastOnNullObject() throws Throwable {
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: add a test case where null is being casted to a refobject

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeGenerator.java [hunk 2]
```diff
@@ -278,6 +279,16 @@ private static void makeValueTypeDefaultValue(ClassWriter cw, String valueName,
 		mv.visitEnd();
 	}
 
+	private static void testCheckCastOnNullObject(ClassWriter cw, String className, String[] fields) {
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeGenerator.java]: add a test case where a non-null value is being casted

File: runtime/vm/BytecodeInterpreter.hpp [hunk 3]
```diff
@@ -7602,24 +7602,25 @@ done:;
 	{
 retry:
 		VM_BytecodeAction rc = EXECUTE_BYTECODE;
+		U_16 index = *(U_16*)(_pc + 1);
+		J9ConstantPool *ramConstantPool = J9_CP_FROM_METHOD(_literals);
+		J9RAMClassRef *ramCPEntry = ((J9RAMClassRef*)ramConstantPool) + index;
+		J9Class* volatile resolvedClass = ramCPEntry->value;
```
Comment (tajila) [inline runtime/vm/BytecodeInterpreter.hpp]: keep the original `castClass` name

**Comment by XxAdi101xX** [comment]
@tajila I've added the test case where a ref class is being casted to null and reverted the variable name change

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 4]
```diff
@@ -1440,6 +1440,28 @@ static public void testDefaultValueInassortedRefWithLongAlignmentArray() throws
 		}
 	}
 
+	/*
+	 * Ensure that casting null to a value type class will throw a null pointer exception 
+	 */
+	@Test(priority=4, expectedExceptions=NullPointerException.class)
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: This can be priority 1

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 5]
```diff
@@ -1440,6 +1440,28 @@ static public void testDefaultValueInassortedRefWithLongAlignmentArray() throws
...
+		String fields[] = {"longField:J"};
+		Class valueClass = ValueTypeGenerator.generateValueClass("TestCheckCastValueTypeOnNull", fields);
+		MethodHandle checkCastValueTypeOnNull = lookup.findStatic(valueClass, "testCheckCastValueTypeOnNull", MethodType.methodType(Object.class));
+		checkCastValueTypeOnNull.invoke();
+	}
+
+	/*
+	 * Ensure that casting null to a reference type class will pass
+	 */
+	@Test(priority=4)
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: same here

Comment (tajila) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: also add a test where a non-NULL VT is being casted

Comment (XxAdi101xX) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: For this test, is it sufficient to just use the DEFAULTVALUE instruction to create a value type and then just checkcast that with the same value type class used to create the object? Or would it be better to create a super class and a base class and try to cast to a super class value type object to a base type one?

Comment (tajila) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java]: The super class of any VT is always j.l.Object so you can just checkcast to the same type.

**Comment by XxAdi101xX** [comment]
@tajila non-null value type casted test has been added and priorities of all the tests have been changed to 1

**Comment by tajila** [comment]
@gacholio please review these changes

File: runtime/vm/BytecodeInterpreter.hpp [hunk 6]
```diff
@@ -7642,6 +7643,17 @@ done:;
 				goto retry;
 			}
 		}
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		else {
+			if ((NULL != castClass) && J9_IS_J9CLASS_VALUETYPE(castClass)) {
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp]: This should be nested ifs that only test castClass once.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 7]
```diff
@@ -7642,6 +7643,17 @@ done:;
...
+				rc = THROW_NPE;
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp]: Is NPE the correct result here?

Comment (tajila) [reply runtime/vm/BytecodeInterpreter.hpp]: Yes according to the <a href="http://cr.openjdk.java.net/~dlsmith/lw2/lw2-20190628/specs/inline-classes-jvms.html#jvms-6.5.checkcast">lw2 spec</a> `if objectref is null and the symbolic reference names a null-free inline class type, the checkcast instruction throws a NullPointerException.`

File: runtime/vm/BytecodeInterpreter.hpp (hunk 6 above)
Comment (XxAdi101xX) [reply runtime/vm/BytecodeInterpreter.hpp]: @gacholio I've rather restructured the ordering of the conditionals so that castClass' is only checked once for null, please take a look at it again

**Comment by gacholio** [comment]
jenkins test sanity xlinux jdk11

**Comment by XxAdi101xX** [comment]
@gacholio I've squashed all my commits into one

**Comment by gacholio** [comment]
jenkins test sanity xlinux jdk11

**Comment by pshipton** [comment]
This isn't compiling on some platforms. It's getting reverted.
```
22:28:26  In file included from BytecodeInterpreter.cpp:102:0:
22:28:26  BytecodeInterpreter.hpp: In member function 'VM_BytecodeAction VM_BytecodeInterpreter::checkcast(UDATA*&, U_8*&)':
22:28:26  BytecodeInterpreter.hpp:7629:1: error: label 'resolve' defined but not used [-Werror=unused-label]
```
